packages/react/src/components/CopyButton/index.tsx

```diff
@@ -0,0 +1,106 @@
+import React, { forwardRef, useState, useCallback, useEffect } from 'react';
+import { createPortal } from 'react-dom';
+import classnames from 'classnames';
+import { ContentNode } from '../../types';
+import Button from '../Button';
+import Offscreen from '../Offscreen';
+import Tooltip from '../Tooltip';
+import Icon from '../Icon';
+import useSharedRef from '../../utils/useSharedRef';
+import copyTextToClipboard from '../../utils/copyTextToClipboard';
+
+type ButtonProps = React.ComponentProps<typeof Button>;
+
+export interface CopyButtonProps
+  extends Omit<ButtonProps, 'onCopy' | 'onClick'> {
+  value: string;
+  variant?: Extract<
+    ButtonProps['variant'],
+    'primary' | 'secondary' | 'tertiary'
+  >;
+  children?: ContentNode;
+  notificationLabel?: ContentNode;
+  hideVisibleLabel?: boolean;
+  tooltipPlacement?: React.ComponentProps<typeof Tooltip>['placement'];
+  onCopy?: (text: string) => void;
+}
+
+const NOTIFICATION_TIMEOUT_MS = 2000;
+
+const CopyButton = forwardRef<HTMLButtonElement, CopyButtonProps>(
+  (
+    {
+      value,
+      variant = 'tertiary',
+      children = 'Copy',
+      notificationLabel = 'Copied',
+      hideVisibleLabel = false,
+      tooltipPlacement = 'auto',
+      onCopy,
+      ...props
+    }: CopyButtonProps,
+    ref
+  ) => {
+    const [copied, setCopied] = useState(false);
+    const copyButtonRef = useSharedRef(ref);
+    const handleClick = useCallback(() => {
+      copyTextToClipboard(value);
+      setCopied(true);
+      if (typeof onCopy === 'function') {
+        onCopy(value);
+      }
+    }, [value, onCopy]);
```

Comment:
We have a pattern in some of our applications where we perform an asynchronous action as part of the copy. For example, if the copy button also updates the sharing visibility of some data, we typically wait until that update happens before confirming to the user that the content has been copied.

I'm not saying that the pattern is ideal, but I'm also trying to think of what that pattern would look like if we tried to incorporate this component into those flows.

Reply:
That's not something that we have an established pattern for in Cauldron/products and may benefit from its own proposal. This implementation is intended to provide the copy mechanism for static synchronous values.